server/src/bam.js

```diff
@@ -3695,18 +3699,47 @@ async function get_gdc_bam(chr, start, stop, gdcFileUUID, bamfilename, req) {
 	const { host, headers } = getGdcDs(req.query.__genomes).getHostHeaders(req.query)
 	headers.compression = false
 	const fullpath = path.join(serverconfig.cachedir_bam, bamfilename)
-	const url = host.rest + '/slicing/view/' + gdcFileUUID + '?region=' + chr + ':' + start + '-' + stop
+	const url = path.join(host.rest, '/slicing/view/', gdcFileUUID + '?region=' + chr + ':' + start + '-' + stop)
 
 	try {
 		const time = Math.round(Date.now() / 1000)
 
 		if (await utils.file_not_exist(fullpath)) {
 			// bam file not found. download
-			await pipelineProm(got.stream(url, { method: 'GET', headers }), fs.createWriteStream(fullpath))
-			// file is supposed to be downloaded
-			if (await utils.file_not_exist(fullpath)) {
-				throw 'BAM file slice is not found after downloading'
+
+			const sourceStream = got.stream(url, { method: 'GET', headers })
+			const writeStream = fs.createWriteStream(fullpath)
+			let totalBytes = 0,
+				tooBigTerminated = false
+			sourceStream.on('data', chunk => {
+				totalBytes += chunk.length
+				if (totalBytes > serverconfig.features.gdcBamCacheMaxSize) {
+					tooBigTerminated = true
+					sourceStream.destroy()
+					writeStream.close()
+					fs.promises.unlink(fullpath)
+				}
+			})
+			try {
+				// not using sourceStream.pipe() so no need to deal with those callbacks as on(close) and on(error)
+				await pipeline(sourceStream, writeStream)
```

Comment:
compared to line 3557, here uses pipeline() but not a.pipe(b), which seems to be the cause of an error that has to be caught and ignored. see if there's a way to fix it.